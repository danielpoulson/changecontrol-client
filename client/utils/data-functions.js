// data-function ver-001 10/06/2017 DP
import sortBy from 'lodash/fp/sortBy';
import compose from 'lodash/fp/compose';
import filter from 'lodash/fp/filter';

const perPage = 15;
const initPage = 1;

export function pagedList(data, page) {
  const _page = page || initPage;
  const offset = (_page - 1) * perPage;
  return data.slice(offset, offset + perPage);
}

export function searchData(data, searchText, sortColumn, columns) {
  const reg1 = new RegExp(`${searchText}.*`, 'i');

  function search(item) {
    if (item[columns[0]].match(reg1) || item[columns[1]].match(reg1) || item[columns[2]].match(reg1)) {
      return true;
    }
    return false;
  }

  let _sortColumn = sortColumn;

  if (typeof sortColumn === 'undefined') {
    _sortColumn = columns[0];
  }

  if (searchText === null) {
    return sortBy(_sortColumn, data);
  }

  const newList = compose(sortBy(_sortColumn), filter(search))(data);

  return newList;
}

export function removeByIndex(data, index) {
  return data.filter(item => item._id !== index);
}

export function getBySourceId(data, id) {
  return data.filter(item => item.SourceId === id);
}
