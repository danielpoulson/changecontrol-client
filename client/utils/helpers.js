/* eslint-disable import/prefer-default-export */
export const fileRegex = /C{2}\d{6}\s[-]\s/;
