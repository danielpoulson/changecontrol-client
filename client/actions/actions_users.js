//SYNC 11/03/2017 DP
import axios from 'axios';
import { serverURL } from '../utils/helpers';

export const GET_USER = 'GET_USER';
export const GET_USERS = 'GET_USERS';
export const RESET_USER = 'RESET_USER';
export const SAVE_USER = 'SAVE_USER';
export const DELETED_USER = 'DELETED_USER';
export const USER_CREATED = 'USER_CREATED';

export function setUsers(users) {
  return { type: GET_USERS, payload: users };
}

export function getUsers() {
  return (dispatch: Function) => {
    axios
      .get(`${serverURL}/api/users/all`)
      .then(response => {
        dispatch(setUsers(response.data));
      })
      .catch(error => {
        console.error('axios error', error); // eslint-disable-line no-console
      });
  };
}

export function addUser(user) {
  return { type: GET_USER, payload: user };
}

export function getUser(id) {
  return (dispatch: Function) => {
    axios
      .get(`${serverURL}/api/users/${id}`)
      .then(response => {
        dispatch(addUser(response.data));
      })
      .catch(error => {
        console.error('axios error', error); // eslint-disable-line no-console
      });
  };
}

export function resetUser() {
  return {
    type: RESET_USER
  };
}

export function addNewUser(fullname) {
  return {
    type: USER_CREATED,
    fullname
  };
}

export function createUser(data) {
  return (dispatch: Function) => {
    axios
      .post(`${serverURL}/api/users`, data)
      .then(() => {
        dispatch(addNewUser(data.fullname));
      })
      .catch(error => {
        console.error('axios error', error); // eslint-disable-line no-console
      });
  };
}

export function saveUser(data) {
  const url = `${serverURL}/api/users/updateuser/${data.username}`;
  axios.put(url, data);

  return {
    type: SAVE_USER,
    data
  };
}

export function savePass(id, password) {
  const url = `${serverURL}/api/users/updatepass/${id}`;
  axios.put(url, { password });

  return {
    type: 'PASSWORD_CHANGED'
  };
}

export function removeUser(fullname) {
  return {
    type: DELETED_USER,
    fullname
  };
}

export function deleteUser(id, fullname) {
  return (dispatch: Function) => {
    axios
      .delete(`${serverURL}/api/users/${id}`)
      .then(() => {
        dispatch(removeUser(fullname));
      })
      .catch(error => {
        console.error('axios error', error); // eslint-disable-line no-console
      });
  };
}
