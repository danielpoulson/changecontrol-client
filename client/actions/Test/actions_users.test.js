import moxios from 'moxios';
import { serverURL } from '../../utils/helpers';
import { addUser, addNewUser, getUser, getUsers, setUsers } from '../actions_users';

const users = [
  'Daniel Poulson',
  'George Saville',
  'Keith Quiney',
  'Matthew Johnston',
  'Paige Finnegan',
  'Patrick Madden'
];

const userData = {
  success: true,
  user: {
    id: '588d97a89be9b78bb608692b',
    fullname: 'Daniel Poulson',
    email: 'danielpoulson@icloud.com',
    username: 'danielp',
    role: 'admin'
  }
};

test('setUsers', () => {
  expect(setUsers(users)).toMatchSnapshot();
});

test('getUsers', (done: Function) => {
  const dispatchMock = jest.fn();
  moxios.withMock(() => {
    getUsers()(dispatchMock);
    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 200,
          response: users
        })
        .then(() => {
          expect(request.url).toEqual(`${serverURL}/api/users/all`);
          expect(dispatchMock).toBeCalledWith(setUsers(users));
          done();
        });
    });
  });
});

test('addUser', () => {
  expect(addUser(userData.user)).toMatchSnapshot();
});

test('getUser', (done: Function) => {
  const dispatchMock = jest.fn();
  moxios.withMock(() => {
    getUser(userData.user.id)(dispatchMock);
    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 200,
          response: userData.user
        })
        .then(() => {
          expect(request.url).toEqual(`${serverURL}/api/users/${userData.user.id}`);
          expect(dispatchMock).toBeCalledWith(addUser(userData.user));
          done();
        });
    });
  });
});

xtest('addNewUser', () => {
  expect(addNewUser(user)).toMatchSnapshot();
});
