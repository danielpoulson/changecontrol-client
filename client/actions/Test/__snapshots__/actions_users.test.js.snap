// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addUser 1`] = `
Object {
  "payload": Object {
    "email": "danielpoulson@icloud.com",
    "fullname": "Daniel Poulson",
    "id": "588d97a89be9b78bb608692b",
    "role": "admin",
    "username": "danielp",
  },
  "type": "GET_USER",
}
`;

exports[`setUsers 1`] = `
Object {
  "payload": Array [
    "Daniel Poulson",
    "George Saville",
    "Keith Quiney",
    "Matthew Johnston",
    "Paige Finnegan",
    "Patrick Madden",
  ],
  "type": "GET_USERS",
}
`;
